source/grid/grid_refinement.cc

```diff
@@ -591,6 +591,98 @@ GridRefinement::refine_and_coarsen_optimize (Triangulation<dim,spacedim> &tria,
   refine (tria, criteria, criteria(tmp[min_arg]));
 }
 
+template<int dim, int spacedim>
+std::pair<unsigned int, double>
+GridRefinement::get_longest_direction(typename Triangulation<dim, spacedim>::active_cell_iterator cell)
+{
+  double max_ratio = 1;
+  unsigned int index = 0;
+
+  for (unsigned int i = 0; i < dim; ++i)
+    {
+      unsigned int ax = i % dim;
+      unsigned int next_ax = (i+1) % dim;
+
+      double ratio =  cell->extent_in_direction(ax) / cell->extent_in_direction(next_ax);
```

Comment:
this algorithm only considers the ratio between successive directions. This covers all possible pairs in 2d and 3d, but it would fail in 4d if we ever go there. I think it would be useful if the algorithm just stored all of the extents in a vector of size dim, and then did a pairwise comparison between all.